backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  AttributeValue,
+  ConditionalCheckFailedException,
+  DynamoDBClient,
+  PutItemCommand,
+  PutItemCommandInput,
+  QueryCommand,
+  QueryCommandInput,
+  QueryCommandOutput,
+} from "@aws-sdk/client-dynamodb";
+import { errorResult, Result, successResult } from "../utils/result";
+import {
+  CreateSessionAttributes,
+  SessionDetails,
+} from "../services/sessionService/sessionService";
+import { NodeHttpHandler } from "@smithy/node-http-handler";
+
+type DynamoDbRecord = Record<string, AttributeValue>;
+
+export class DynamoDbAdapter {
+  private readonly tableName: string;
+  private readonly dynamoDbClient = new DynamoDBClient({
+    region: process.env.REGION,
+    maxAttempts: 2,
+    requestHandler: new NodeHttpHandler({
+      connectionTimeout: 29000,
+      requestTimeout: 29000,
+    }),
+  });
+
+  constructor(tableName: string) {
+    this.tableName = tableName;
+  }
+
+  async readSessionId(
+    subjectIdentifier: string,
+  ): Promise<Result<string | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: "ASYNC_AUTH_SESSION_CREATED" },
+        ":currentTimeInSeconds": {
+          N: currentTimeInSeconds.toString(),
+        },
+      },
+      ProjectionExpression: "sessionId",
+      Limit: 1,
```

Comment:
Does this retrieve the most recent if there are two sessions that fit this criteria (that's what we should be aiming for I think)

Reply:
I did not change the part of the original query that was meant to be doing that (i.e. retrieving the most recent session). I didn't test it because I expected that if that code was merged, it was tested for this scenario and I could just re-use it and adapt the query to return the session values I wanted.

I think that setting `ScanIndexForward` to `false` means that it will return the most recent added session (this is part of the code that's in `main`): 

`If ScanIndexForward is true, DynamoDB returns the results in the order in which they are stored (by sort key value). This is the default behavior. If ScanIndexForward is false, DynamoDB reads the results in reverse order by sort key value, and then returns the results to the client.
`

However, I can do a manual test as well...

Reply:
I'm just thinking about this more... as we've updated the POST /credential endpoint to only create a session if one isn't available, I think we're fine here.

I agree with your analysis in any case